src/main/java/spoon/reflect/path/impl/CtRolePathElement.java

```diff
@@ -106,14 +55,57 @@ public CtRole getRole() {
 
 	@Override
 	public String toString() {
-		return STRING + role.toString() + getParamString();
+		return STRING + getRole().toString() + getParamString();
+	}
+
+	public CtElement getFromSet(Set set, String name) throws CtPathException {
+		for (Object o: set) {
+			if (o instanceof CtNamedElement) {
+				if (((CtNamedElement) o).getSimpleName().equals(name)) {
+					return (CtElement) o;
+				}
+			} else if (o instanceof CtReference) {
+				if (((CtReference) o).getSimpleName().equals(name)) {
+					return (CtElement) o;
+				}
+			} else {
+				throw new CtPathException();
+			}
+		}
+		throw new CtPathException();
 	}
 
 	@Override
 	public Collection<CtElement> getElements(Collection<CtElement> roots) {
-		RoleVisitor visitor = new RoleVisitor();
-		visitor.scan(roots);
-		return visitor.matchs;
+		Collection<CtElement> matchs = new LinkedList<>();
+		for (CtElement root : roots) {
+			try {
+				if (root.getValueByRole(getRole()) instanceof List) {
+					if (getArguments().containsKey("index")) {
+						int index = Integer.parseInt(getArguments().get("index"));
+						matchs.add((CtElement) ((List) root.getValueByRole(getRole())).get(index));
+					}
+				} else if (root.getValueByRole(getRole()) instanceof Set) {
+					if (getArguments().containsKey("name")) {
+						String name = getArguments().get("name");
+						try {
+							matchs.add(getFromSet(root.getValueByRole(getRole()), name));
+						} catch (CtPathException e) {
+							//System.err.println("[ERROR] Element not found for name: " + name);
+							//No element found for name.
+						}
+					}
+				} else if (root.getValueByRole(getRole()) instanceof Map) {
+					if (getArguments().containsKey("key")) {
+						String name = getArguments().get("key");
+						matchs.add((CtElement) ((Map) root.getValueByRole(getRole())).get(name));
+					}
+				} else {
+					CtElement el = root.getValueByRole(getRole());
+					matchs.add(el);
+				}
+			} catch (SpoonException e) { } //When no element are found for a given role, return empty list.
```

Comment:
I guess one alternative would be adding a method `hasChildFor(CtRole role)` to `CtElement`.

Reply:
It is already there. See `RoleHandlerHelper#getOptionalRoleHandler`
Have a look also at `RoleHandler#getContainerKind`, which might be used instead of `root.getValueByRole(getRole()) instanceof Set` and similar

Reply:
@pvojtechovsky Thanks for the tip.
Problem solved.